shim/src/proxy.js

```diff
@@ -0,0 +1,68 @@
+import {
+  defineProperty,
+  deleteProperty,
+  getOwnPropertyDescriptor,
+  getOwnPropertyNames
+} from './commons';
+
+// this flag allow us to determine if the eval() call is a controlled eval done by the realm's code
+// or if it is user-land invocation, so we can react differently.
+let isInternalEvaluation = false;
+
+export function setInternalEvaluation() {
+  isInternalEvaluation = true;
+}
+
+export function resetInternalEvaluation() {
+  isInternalEvaluation = false;
+}
+
+export const proxyHandler = {
+  get(sandbox, propName) {
+    if (propName === 'eval' && isInternalEvaluation) {
+      resetInternalEvaluation();
+      return sandbox.confinedWindow.eval;
+    }
+    return sandbox.globalObject[propName];
+  },
+  set(sandbox, propName, newValue) {
+    sandbox.globalObject[propName] = newValue;
+    return true;
+  },
+  defineProperty(sandbox, propName, descriptor) {
+    defineProperty(sandbox.globalObject, propName, descriptor);
+    return true;
+  },
+  deleteProperty(sandbox, propName) {
+    return deleteProperty(sandbox.globalObject, propName);
+  },
+  has(sandbox, propName) {
+    if (propName === 'eval' && isInternalEvaluation) {
+      return true;
+    }
+    if (propName in sandbox.globalObject) {
+      return true;
+    } else if (propName in sandbox.confinedWindow) {
+      throw new ReferenceError(
+        `${propName} is not defined. If you are using typeof ${propName}, you can change your program to use typeof global.${propName} instead`
+      );
+    }
+    return false;
+  },
+  ownKeys(sandbox) {
+    return getOwnPropertyNames(sandbox.globalObject);
+  },
+  getOwnPropertyDescriptor(sandbox, propName) {
+    return getOwnPropertyDescriptor(sandbox.globalObject, propName);
```

Comment:
doesn't check isInternalEvaluation. Can't it leak the true unsafe evaluator?


I think the answer to both of the questions above may be that this proxy is only used in a with and can never escape, in which case we only care about the traps that can happen from this with-usage. Is this true? Would be good to know what code behavior in with-body triggers what traps. Other traps should error rather than default.


I am in general confused about why the sandbox is used as the target object. For example, the extensibility of the target should reflect the extensibility of the emulated global, which I think is sandbox.globalObject, not sandbox.

Reply:
Original code. Removed.